application/controllers/card_logic.py

```diff
@@ -69,6 +69,178 @@ def process_card(card, player, is_discarded, for_wonder, from_discard_pile=None,
     return True
 
 
+def trade(card, player, trade=True):
+    # Get cards (yellow and non-base resource wonder cards are not used for trade)
+    cards = [c for c in get_cards(player=player, history=True)]
+    left_cards = [c for c in get_cards(player=get_player(player.left_id), history=True) if not c.colour == 'yellow' and not c.colour == 'wonder' or c.colour == 'wonder' and c.name.endswith('_0')]
+    right_cards = [c for c in get_cards(player=get_player(player.right_id), history=True) if not c.colour == 'yellow' and not c.colour == 'wonder' or c.colour == 'wonder' and c.name.endswith('_0')]
+
+    # Process cards into queue
+    needed = get_card_requirements(card)
+    card_options = []
+    create_list(cards, card_options, [0, 0], 'player')
+    create_list(left_cards, card_options,
+                [1 if player.left_cheap_trade else 2, 1 if player.advanced_cheap_trade else 2], 'left')
+    create_list(right_cards, card_options,
+                [1 if player.right_cheap_trade else 2, 1 if player.advanced_cheap_trade else 2], 'right')
+    card_options = sorted(card_options, key=lambda k: k[list(k.keys())[0]][1] if k[list(k.keys())[0]] != 'Alternating' else k[list(k.keys())[1]][1])
+
+    # Discard cards not needed
+    prune_cards(needed, card_options)
+
+    # Allocate resources
+    stats = []
+    assign_cards(card_options, needed, 0, stats, {'left': {'cost': 0}, 'right': {'cost': 0}, 'possible': False})
+
+    # Returns best option if successful or default if check fails
+    if stats:
+        return stats[0]
+    else:
+        return {'left': {'cost': 0}, 'right': {'cost': 0}, 'possible': False}
+
+
+def assign_cards(card_options, needed, current_price, stats, tempstats):
+    while card_options and any(needed):
+        # If this branch has reached the price of the best option, prune
+        if stats and tempstats['left']['cost'] + tempstats['right']['cost'] >= stats[0]['left']['cost'] + stats[0]['right']['cost']:
+            return
+        print("Needed:", needed)
+
+        # Choose next card to play. Favour non RA cards where price is equal
+        c = None
+        index = 0
+        while not c:
+            if len(card_options) > index:
+                temp = card_options[index]
+                for i in temp.keys():
+                    if i != 'Alternating' and temp[i][1] > current_price:
+                        c = card_options.pop(0)
+                        current_price = c[i][1]
+                        break
+                    elif i != 'Alternating' and temp[i][1] == current_price:
+                        c = card_options.pop(index)
+                        break
+                    else:
```

Comment:
Is there never another possibility? We should probably have something to catch a potential future error?

Reply:
There's never any other possibilities afaik, as the 'cards' get created by get_card_benefits, which always gives an Alternating key with a boolean, and then everything else is a key with an array.

def get_card_benefits(card, prices, player):
    temp = {}
    if card.giveStone:
        temp['Stone'] = [card.giveStone, prices[0], player]
    if card.giveBrick:
        temp['Brick'] = [card.giveBrick, prices[0], player]
    if card.giveOre:
        temp['Ore'] = [card.giveOre, prices[0], player]
    if card.giveWood:
        temp['Wood'] = [card.giveWood, prices[0], player]
    if card.giveGlass:
        temp['Glass'] = [card.giveGlass, prices[1], player]
    if card.givePaper:
        temp['Paper'] = [card.givePaper, prices[1], player]
    if card.giveCloth:
        temp['Cloth'] = [card.giveCloth, prices[1], player]
    if card.resourceAlternating:
        temp['Alternating'] = True
    else:
        temp['Alternating'] = False
    return temp